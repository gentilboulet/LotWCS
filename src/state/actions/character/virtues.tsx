import { createAction } from "typesafe-actions";
import { ICost } from "../../models/character/costs";
import { historyMetaCreator } from "../meta";

export const increase = createAction(
  "virtues/INCREASE",
  (name: string, value: number, cost: ICost) => ({ cost, name, value }),
  historyMetaCreator,
)();
