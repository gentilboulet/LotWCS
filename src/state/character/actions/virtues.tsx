import { createAction } from "typesafe-actions";
import { ICost } from "../models/costs";

export const increase = createAction(
  "virtues/INCREASE",
  action => (name: string, value: number, cost: ICost) =>
    action({ name, value, cost }),
);
