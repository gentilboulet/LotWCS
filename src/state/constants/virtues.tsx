export const VIRTUE_INCREASE = 'VIRTUE_INCREASE';
export type VIRTUE_INCREASE = typeof VIRTUE_INCREASE;

export const VIRTUE_OPEN = 'VIRTUE_OPEN';
export type VIRTUE_OPEN = typeof VIRTUE_OPEN;
