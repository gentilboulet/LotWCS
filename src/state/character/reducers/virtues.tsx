import produce from "immer";
import { ActionType, getType } from "typesafe-actions";

import * as actions from "../actions/virtues";
import { ICharacterState } from "../models/type";

import { applyCost } from "../models/costs";
import { increase } from "../models/virtues";

export function virtuesReducer(
  baseState: ICharacterState,
  action: ActionType<typeof actions>
): ICharacterState {
  switch (action.type) {
    case getType(actions.increase):
      return produce(baseState, state => {
        applyCost(state, action.payload.cost);
        increase(state.virtues, action.payload.name, action.payload.value);
      });
    default:
  }
  return baseState;
}
