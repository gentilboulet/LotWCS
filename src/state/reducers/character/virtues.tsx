import { produce, Draft } from "immer";
import { ActionType, getType } from "typesafe-actions";

import * as actions from "../../actions/character/virtues";
import { increase, TVirtuesState } from "../../models/character/virtues";

export const virtuesReducer = produce(
  (draft: Draft<TVirtuesState>, action: ActionType<typeof actions>) => {
    switch (action.type) {
      case getType(actions.increase):
        increase(draft, action.payload.name, action.payload.value);
        break;
      default:
    }
  },
);
