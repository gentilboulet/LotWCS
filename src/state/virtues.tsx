import { IDataVirtue, IDataVirtueType, virtues } from 'data/virtues';
import { canPayCost, getCostVirtue } from 'state/costs';
import { IStoreState } from 'state/type';

export interface IVirtueState {
    name: string;
    value: number;
    type: IDataVirtueType;
}

export type TVirtuesState = IVirtueState[];

export function createState(): TVirtuesState {
  return virtues.map((virtue: IDataVirtue) => {
    return { name: virtue.name, value: 0, type: virtue.type };
  });
}

export function add(state: TVirtuesState, name: string, type: IDataVirtueType, value: number): void {
  if(isVirtuePresent(state, name)) { throw new Error('Internal Error : virtue already present ' + name); }
  state.push({ name, value, type });
}

export function increase(state: TVirtuesState, name: string, value: number): void {
  if(! isVirtuePresent(state, name)) { throw new Error('Internal Error : unknwon new virtue ' + name); }
  state.forEach(virtue => {
    if(virtue.name === name) { virtue.value+= value; }
  })
}

export function isVirtuePresent(state: TVirtuesState, name: string): boolean {
  return state.findIndex(virtue => name === virtue.name) !== -1;
}

export function canBuyVirtue(state: IStoreState, name: string): boolean {
  const findIndex = state.virtues.findIndex(v => name === v.name);
  if( findIndex === -1 ) { return false; }
  const virtue = state.virtues[findIndex];
  if( (virtue.value+1) > 5 )  { return false; }

  const cost = getCostVirtue(state, name);
  return canPayCost(state, cost);
}
