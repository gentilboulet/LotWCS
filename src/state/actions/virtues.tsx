import * as constants from 'constants/virtues';
import * as dataVirtues from 'data/virtues';

import { ICost } from 'costs/types';

export interface IVirtueIncrease {
  type: constants.VIRTUE_INCREASE;
  name: string;
  virtueType: dataVirtues.IDataVirtueType;
  value: number;
  cost: ICost;
}

export type IVirtueAction = IVirtueIncrease;

export function increase(name: string, virtueType: dataVirtues.IDataVirtueType, value: number, cost: ICost): IVirtueIncrease {
  return { type: constants.VIRTUE_INCREASE, name, virtueType, value, cost };
}
