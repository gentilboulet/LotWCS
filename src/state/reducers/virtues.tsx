import { IVirtueAction } from 'state/actions/virtues';
import * as constants from 'state/constants/virtues';
import { IStoreState } from 'state/types';

import { pushToHistory } from 'state/history';
import * as virtues from 'state/virtues';

export function virtuesReducer(oldState: IStoreState, action: IVirtueAction): IStoreState {
  switch (action.type) {
    case constants.VIRTUE_INCREASE:
      return oldState.withMutations(state => {
        if (virtues.getVirtueIndex(state, name) === -1) {
          virtues.addVirtue(state, action.name, action.virtueType);
        }
        virtues.increaseVirtue(state, action.name, action.value);
        pushToHistory(state, action);
      });
    default:
  }
  return oldState;
}
