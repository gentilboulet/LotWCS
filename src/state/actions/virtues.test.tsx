import * as dataVirtues from 'data/virtues';
import * as actions from './virtues';

describe('Testing virtue action creator', () => {
  const noCost = {
    destiny: 0,
    discountIdx: -1, discountNewValue: 0,
    entanglement: 0
  };

  dataVirtues.virtues.map((data: dataVirtues.IDataVirtue) => {
    it('should create a virtue increase action with an existing virtue', () => {
      const name = data.name;
      const a = actions.increase(name, 10, noCost);
      expect( a.name ).toBe(name);
      expect( a.value ).toBe(10);
      expect( a ).toMatchSnapshot();
    });
  });

  it('should create a virtue increase action with a non existing virtue', () => {
    const name = 'New virtue name';
    const a = actions.increase(name, 3, noCost);
    expect( a.name ).toBe(name);
    expect( a.value ).toBe(3);
    expect( a ).toMatchSnapshot();
  });
});
