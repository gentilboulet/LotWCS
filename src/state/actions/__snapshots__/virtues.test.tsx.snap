// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing virtue action creator should create a virtue increase action with a non existing virtue 1`] = `
Object {
  "cost": Object {
    "destiny": 0,
    "discountIdx": -1,
    "discountNewValue": 0,
    "entanglement": 0,
  },
  "name": "New virtue name",
  "type": "VIRTUE_INCREASE",
  "value": 3,
  "virtueType": "VIRTUE_SELFISH",
}
`;

exports[`Testing virtue action creator should create a virtue increase action with an existing virtue 1`] = `
Object {
  "cost": Object {
    "destiny": 0,
    "discountIdx": -1,
    "discountNewValue": 0,
    "entanglement": 0,
  },
  "name": "Honor",
  "type": "VIRTUE_INCREASE",
  "value": 10,
  "virtueType": "VIRTUE_CHIVALROUS",
}
`;

exports[`Testing virtue action creator should create a virtue increase action with an existing virtue 2`] = `
Object {
  "cost": Object {
    "destiny": 0,
    "discountIdx": -1,
    "discountNewValue": 0,
    "entanglement": 0,
  },
  "name": "Benevolence",
  "type": "VIRTUE_INCREASE",
  "value": 10,
  "virtueType": "VIRTUE_CHIVALROUS",
}
`;

exports[`Testing virtue action creator should create a virtue increase action with an existing virtue 3`] = `
Object {
  "cost": Object {
    "destiny": 0,
    "discountIdx": -1,
    "discountNewValue": 0,
    "entanglement": 0,
  },
  "name": "Righteousness",
  "type": "VIRTUE_INCREASE",
  "value": 10,
  "virtueType": "VIRTUE_CHIVALROUS",
}
`;

exports[`Testing virtue action creator should create a virtue increase action with an existing virtue 4`] = `
Object {
  "cost": Object {
    "destiny": 0,
    "discountIdx": -1,
    "discountNewValue": 0,
    "entanglement": 0,
  },
  "name": "Loyalty",
  "type": "VIRTUE_INCREASE",
  "value": 10,
  "virtueType": "VIRTUE_CHIVALROUS",
}
`;

exports[`Testing virtue action creator should create a virtue increase action with an existing virtue 5`] = `
Object {
  "cost": Object {
    "destiny": 0,
    "discountIdx": -1,
    "discountNewValue": 0,
    "entanglement": 0,
  },
  "name": "Force",
  "type": "VIRTUE_INCREASE",
  "value": 10,
  "virtueType": "VIRTUE_CHIVALROUS",
}
`;

exports[`Testing virtue action creator should create a virtue increase action with an existing virtue 6`] = `
Object {
  "cost": Object {
    "destiny": 0,
    "discountIdx": -1,
    "discountNewValue": 0,
    "entanglement": 0,
  },
  "name": "Revenge",
  "type": "VIRTUE_INCREASE",
  "value": 10,
  "virtueType": "VIRTUE_SELFISH",
}
`;

exports[`Testing virtue action creator should create a virtue increase action with an existing virtue 7`] = `
Object {
  "cost": Object {
    "destiny": 0,
    "discountIdx": -1,
    "discountNewValue": 0,
    "entanglement": 0,
  },
  "name": "Individualism",
  "type": "VIRTUE_INCREASE",
  "value": 10,
  "virtueType": "VIRTUE_SELFISH",
}
`;

exports[`Testing virtue action creator should create a virtue increase action with an existing virtue 8`] = `
Object {
  "cost": Object {
    "destiny": 0,
    "discountIdx": -1,
    "discountNewValue": 0,
    "entanglement": 0,
  },
  "name": "Obsession",
  "type": "VIRTUE_INCREASE",
  "value": 10,
  "virtueType": "VIRTUE_SELFISH",
}
`;

exports[`Testing virtue action creator should create a virtue increase action with an existing virtue 9`] = `
Object {
  "cost": Object {
    "destiny": 0,
    "discountIdx": -1,
    "discountNewValue": 0,
    "entanglement": 0,
  },
  "name": "Ruthlessness",
  "type": "VIRTUE_INCREASE",
  "value": 10,
  "virtueType": "VIRTUE_SELFISH",
}
`;

exports[`Testing virtue action creator should create a virtue increase action with an existing virtue 10`] = `
Object {
  "cost": Object {
    "destiny": 0,
    "discountIdx": -1,
    "discountNewValue": 0,
    "entanglement": 0,
  },
  "name": "Ferocity",
  "type": "VIRTUE_INCREASE",
  "value": 10,
  "virtueType": "VIRTUE_SELFISH",
}
`;
